include/cxxeffect/task.hpp

```diff
@@ -0,0 +1,53 @@
+#pragma once
+#include <functional>
+#include <future>
+
+namespace eff {
+    template <class F, class T>
+    using flatmap_t = decltype(std::declval<F>()(std::declval<T>()).unsafeRunSync());
+    template <class F, class T>
+    using map_t = decltype(std::declval<F>()(std::declval<T>()));
+    template <class T>
+    class task {
+        std::function<T()> thunk;
+       public:
+        task() = default;
+        task(task const&) = default;
+        task(task&&) = default;
+        task(std::function<T()> thunk) : thunk(std::move(thunk)) {}
+
+        static task pure(T value) {
+            return task([=]() { return value; });
+        }
+        static auto delay(std::function<T()> thunk) {
+            return task(thunk);
+        }
+
+        template <class F>
+        auto map(F func) const -> task<map_t<F, T>> {
+            return {
+                    [f = std::move(func), input = *this]() {
```

Comment:
I didn't even know you could rewrite captures like this. I love it

Reply:
Lamdas can be quite nice 🥰